httpx/cert_reload.go

```diff
@@ -0,0 +1,159 @@
+package httpx
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"path"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ory/x/logrusx"
+)
+
+type (
+	CertLocation struct {
+		CertPath string
+		KeyPath  string
+	}
+
+	CertLoadFunc func() ([]tls.Certificate, error)
+
+	CertProvider struct {
+		certs   []tls.Certificate
+		mu      sync.Mutex
+		watcher *fsnotify.Watcher
+
+		certLoad     CertLoadFunc
+		certLocation CertLocation
+		logger       *logrusx.Logger
+	}
+)
+
+//NewCertProvider creates a tls.Certificate provider that reloads certificates on fs change
+//certs must not be empty
+func NewCertProvider(certs []tls.Certificate, certLocation CertLocation, certLoad CertLoadFunc, l *logrusx.Logger) *CertProvider {
```

Comment:
Personally, I find this interface a little bit confusing. We give a filesystem location to the function to watch, but we never actually read from the file (instead deferring to the `certLoad` callback). Do we need this complexity? Same with the initial `certs` argument. I think the interface would be better if it just accepted the location and did all reading internally.

Reply:
I'm not familiar with the Kratos and Hydro codebases, so I can't comment really 😁 . Have you taken a look at #507 though? It adds this functionality to `x/tlsx` instead of `x/httpx`.

Reply:
No worry I can move it to tlsx if you wish but that does not answer my question as this will be used for Hydra and Kratos then.

@aeneasr requested me to merge the common things here in x. The question still remains, should we rewrite TLS confi loading in a unified way for Kratos and Hydra ?

Reply:
> I guess there's another one question buried here: How do we deal with the config value changing? So let's say `serve.tls.cert=file://abcd.pem` changes to `serve.tls.cert=file://not-abcd.pem`. How would we reload that with the current approach?


We should consider separating the notions of config hot reloading and TLS certificate reloading from an implementational POV. It's simpler to reason about a package/function which
1.  reads certs from a specific filesystem path
2.  reloads the files when they change
3.  allows the background reload to be cancelled

than it is to reason about a package which additionally also
4.  reads certificates from a base64 string (but only sometimes)
5.  generates a self-signed certificate (but only sometimes)
6.  changes its behavior completely during its lifetime

My proposal is to keep the interface of the cert reloader minimal. When the underlying config in Hydra/Kratos changes ("`serve.tls.cert=file://abcd.pem` changes to `serve.tls.cert=file://not-abcd.pem`"), the code in Hydra/Kratos should detect that, cancel the old TLS cert reloader, and spawn a new one pointing to the new path.

With a little bit of glue code in `github.com/ory/hydra/config` it should still be possible to do hot-reloading without closing the listening socket.

Reply:
My current implementation expect that on config reload you call LoadCertificates to perform the changes.

@aeneasr Please confirm that "serve.tls.cert=file://not-abcd.pem" is already handled by config handlers in hydra / kratos and does not impact.